lib/index.js

```diff
@@ -107,23 +115,48 @@ api.set = async ({
   } else if(type === 'nonce') {
     // FIXME: may need to allow up to N tokens, one per `clientId` if
     // `clientId` is set
-
-    // if there is an existing, unexpired nonce, reuse it, do not generate
-    // a new one to avoid DoS attack of constantly running bcrypt algorithm
+    const cfg = config['authn-token'];
+    const {maxNonceCount} = cfg.nonce.defaults;
+    let existingTokens;
     try {
-      const existingToken = await api.get({actor: null, account, email, type});
-      // keep existing token if it hasn't expired; note that there is no
-      // `challenge` to return because it is not known in the database
-      if(existingToken && new Date() < new Date(existingToken.expires)) {
-        return {type};
-      }
+      // check if nonce exists
+      existingTokens = await api.getAll({
+        actor: null, account, email, type, filterExpiredTokens: false
+      });
```

Comment:
(Not sure where to leave this comment)  There are likely race conditions in this code, depending how it can get called.  How critical is it that they be better handled?  Testing may be challenging.  I think one outcome would be allowing > max nonce tokens.  That may be something that can be intentionally triggered by sending many concurrent requests:

- Concurrent calls to `set()` could each calculate enough space was available from `getAll` and then each add tokens resulting in > max tokens. (Might be easy to trigger.)
- Concurrent calls to `set()` could each be in the expiration loop and think they removed a token leaving space to add a new one, resulting in > max tokens. (Probably harder/rare to trigger.)

There might be variations of these paths, or some that don't allow a set when it should be allowed.  It's a bit difficult to reason through these cases.

A small improvement for the expiration loop issue might be for `remove()` to return if the token was removed from the db, and only count that as a removal.

If these issues are a concern, perhaps a modified design might help limit issues.  Only consider the most recent maxNonceCount unexpired tokens to be valid.  In `set()`, add a sortable field, then when removing tokens, remove both expired, and any older than most recent maxNonceCount ones.  If this was a real problem could add extra logic to `get` calls to filter all but max allowable count of tokens.  Adding a proper sortable field may be difficult in a distributed system.  Maybe just mongo auto _id would be fine in this case.  I'm unsure if any of this is a enough of a concern to take the performance and complexity hit.

@dlongley Thoughts?

Reply:
Yeah, missed this before. I think what should happen should be that the `update` should include the previous values in its query so that there's no match if there has been a change while we were here in the application computing things. That should be simple enough -- we have done this pattern in other places with success.

Reply:
I'm not sure we've done that pattern when there are multiple rows of data?  Seems like there would still be race conditions unless we started using transactions.

Reply:
My understanding is that there aren't multiple rows of data here. Currently, multiple tokens are stored in a single account record. I do see that `email` could be used here in the query for multiple accounts -- bedrock-account allows for more than one account per email at present -- so, there seems to be a problem here with that and how the limits are applied. The limits for nonces should be applied *per account*, not per "email". So I think this area still needs a little work to account for that. However, when updating a particular account, we should be able to compare the tokens in the mongo update query against what was ready out to ensure we only make changes if no intervening/concurrent update occurred.

Reply:
@JSAssassin -- it looks like the `getAll` query used here could be using `email` -- which is problematic as the tokens returned could be for multiple accounts. We want to resolve the email to N accounts and process them each separately so that the max nonce limitation applies *per* account. Then, when we attempt to update each account, we should make sure that the mongo update query includes whatever the previous token information was to ensure the update is not applied if it changed concurrently. When that happens, we can detect that the update was not applied and we should just return a 409 conflict error for simplicity. If we discover that's messy in the future we can try to do something more elegant.

Reply:
> Then, when we attempt to update each account, we should make sure that the mongo update query includes whatever the previous token information was to ensure the update is not applied if it changed concurrently. 

@dlongley Is there an example that I can follow?

Reply:
Oops, nevermind that multiple row comment. ;-)

Checking previous data seems tricky to get right due to the potential concurrent `remove()` calls.  Perhaps should just rebuild the whole array with filtered out expired tokens and the added new one then use $set instead of $push.  Then you would have both the old and new state to use in the update.  I'm forgetting where else we do this, do we just return 409 or do a retry loop first?

Reply:
> I'm forgetting where else we do this, do we just return 409 or do a retry loop first?

We do both in different places depending on the use case. Here we should just return 409 for simplicity, IMO. Feel free to argue otherwise.

Reply:
We're going to make emails unique over in bedrock-account to simplify code here and elsewhere (and to remove a feature that is rarely used and creates unnecessary complexity). Once that's done, the code here will simplify based on the assumption that we're only ever working with a single record (since email maps to exactly 1 account).

Reply:
@davidlehn -- +1 to doing a single update at the end that does `$set` (and compares against the previous record) after removing all expired tokens. If it fails to update a record in the database, it should result in a 409 conflict.

Reply:
@davidlehn @dlongley I have updated the PR but I am not sure how I can compare against previous record.

Reply:
@JSAssassin, so to compare against the previous account record, we're going to need to include the old information in the update `query`.

I see that we have a `filterExpiredTokens: false` flag being passed to `getAll` here. Assuming we don't change the order or appearance of the tokens we return in any way from `getAll`, then we can just include the old tokens in the `query` below -- and that will cause the comparison to occur.